dice-where-downloader-lib/src/main/java/technology/dice/dicewhere/downloader/stream/StreamWithMD5Decorator.java

```diff
@@ -1,41 +1,82 @@
 package technology.dice.dicewhere.downloader.stream;
 
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.security.DigestInputStream;
 import java.security.MessageDigest;
 import java.security.NoSuchAlgorithmException;
+import java.util.Optional;
 import javax.xml.bind.annotation.adapters.HexBinaryAdapter;
 import technology.dice.dicewhere.downloader.md5.MD5Checksum;
 
 public class StreamWithMD5Decorator extends InputStream {
 
+  private static final int BUFFER_SIZE = 8192;
+  private static final HexBinaryAdapter HEX_BINARY_ADAPTER = new HexBinaryAdapter();
+
   private final MessageDigest md5;
-  DigestInputStream inputStream;
+  private final DigestInputStream originalStream;
+
+  private Optional<InputStream> duplicatedStream = Optional.empty();
+  private Optional<MD5Checksum> md5Checksum = Optional.empty();
 
-  private StreamWithMD5Decorator(DigestInputStream inputStream, MessageDigest md5) {
-    this.inputStream = inputStream;
+  private StreamWithMD5Decorator(DigestInputStream originalStream, MessageDigest md5)
+      throws IOException {
+    this.originalStream = originalStream;
     this.md5 = md5;
+    consumeStream();
   }
 
-  public static StreamWithMD5Decorator of(InputStream inputStream) throws NoSuchAlgorithmException {
+  public static StreamWithMD5Decorator of(InputStream inputStream)
+      throws NoSuchAlgorithmException, IOException {
     MessageDigest md5 = MessageDigest.getInstance("MD5");
     DigestInputStream dis = new DigestInputStream(inputStream, md5);
     return new StreamWithMD5Decorator(dis, md5);
   }
 
+  private void consumeStream() throws IOException {
+    ByteArrayOutputStream tempBufferStream = new ByteArrayOutputStream();
+    byte[] data = new byte[BUFFER_SIZE];
+    int bytesRead;
+    while ((bytesRead = originalStream.read(data)) != -1) {
+      tempBufferStream.write(data, 0, bytesRead);
+    }
+    duplicatedStream = Optional.of(new ByteArrayInputStream(tempBufferStream.toByteArray()));
+  }
+
   public MD5Checksum md5() {
-    String hex = (new HexBinaryAdapter()).marshal(this.md5.digest());
-    return MD5Checksum.of(hex);
+    if (duplicatedStream.isEmpty()) {
```

Comment:
aren't we consuming stream on the constructor? is this possible to happen?
if not, instead of throwing an exception, wouldn't be possible to make sure we consume the stream here?